ql/src/java/org/apache/hadoop/hive/ql/cleanup/EventualCleanupService.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.cleanup;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class EventualCleanupService implements CleanupService {
+  private final int threadCount;
+  private final int queueSize;
+  private final ThreadFactory factory;
+  private final Logger LOG = LoggerFactory.getLogger(EventualCleanupService.class.getName());
+  private final AtomicBoolean isRunning = new AtomicBoolean(true);
+  private final BlockingQueue<AsyncDeleteAction> deleteActions;
+  private ExecutorService cleanerExecutorService;
+
+  public EventualCleanupService(int threadCount, int queueSize) {
+    if (queueSize < threadCount) {
+      throw new IllegalArgumentException("Queue size should be greater or equal to thread count. Queue size: "
+          + queueSize + ", thread count: " + threadCount);
+    }
+    this.factory = new ThreadFactoryBuilder().setDaemon(true).setNameFormat("EventualCleanupService thread %d").build();
+    this.threadCount = threadCount;
+    this.queueSize = queueSize;
+    this.deleteActions = new LinkedBlockingQueue<>(queueSize);
+  }
+
+  @Override
+  public synchronized void start() {
+    if (cleanerExecutorService != null) {
+      LOG.debug("EventualCleanupService is already running.");
+      return;
+    }
+    cleanerExecutorService = Executors.newFixedThreadPool(threadCount, factory);
+    for (int i = 0; i < threadCount; i++) {
+      cleanerExecutorService.submit(new CleanupRunnable());
```

Comment:
I might be wrong but:
* I think these threads will be there all the time - even if there is no work to be done (let's use `N` for this)
* it seems to me that every session will have these cleaner threads for them (current session count: `M`)

...so there will be `N*M` threads running all the time?

Reply:
CleanupService is one per HiveServer, not per session. So there will be N threads running all the time (actually blocked on blocking queue when there is no cleanup to be done, so not scheduled when unncessary).